### PR TITLE
Yet Yet another batch of decomp-related fixes

### PR DESCRIPTION
**Pull request by astralcai**
**Context:**
More decomposition related fixes.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-109010]
[sc-111284]

### CODE REVIEW

File: pennylane/math/decomposition.py
Comment (astralcai) [inline pennylane/math/decomposition.py:1]: Getting rid of all `math.where` because they are discontinuities that causes `nan` gradients sometimes.

File: pennylane/ops/op_math/controlled_ops.py [hunk 1]
```diff
@@ -251,12 +251,41 @@ def _controlled(self, wire):
...
+    the graph finds the general decomposition rule of applying control to the decomposition
+    of the base QubitUnitary."""
+    num_target_wires = len(wires) - len(control_wires)
+    qml.ops.Controlled(
+        qml.QubitUnitary(U, wires=wires[-num_target_wires:]),
+        control_wires=control_wires,
+        control_values=control_values,
+        work_wires=work_wires,
+        work_wire_type=work_wire_type,
+    )
```
Comment (astralcai) [inline pennylane/ops/op_math/controlled_ops.py:279]: The current logic in the decomposition graph for controlled operators is that, we only consider the basic decomposition (of applying control to the decomposition of the target operator) for general controlled operators. For special controlled ops like `CNOT`, `CRX`, `CSWAP`, etc., we don't do that. Here `ControlledQubitUnitary` is a special case where the special decomposition rules for `ControlledQubitUnitary` only go up to two target wires, while a `ControlledQubitUnitary` could have any number of target wires.

This is the least invasive fix in my opinion, simply converting a special `ControlledQubitUnitary` class to a general controlled `QubitUnitary`, allowing the decomposition to be applied by first decomposing its target op (the `QubitUnitary`, and then applying the control).

File: pennylane/transforms/decompose.py [hunk 2]
```diff
@@ -227,6 +227,7 @@ def eval(self, jaxpr: jax.extend.core.Jaxpr, consts: Sequence, *args) -> list:
                     operations = collector.state["ops"]
 
                 if operations:
+                    operations = [op for op in operations if not self.stopping_condition(op)]
```
Comment (astralcai) [inline pennylane/transforms/decompose.py:230]: We don't try to build the graph out for an operator if the operator (although not in the gate set) already meets the stopping condition.

File: pennylane/devices/preprocess.py [hunk 3]
```diff
@@ -374,8 +373,8 @@ def decompose(  # pylint: disable = too-many-positional-arguments
 
     graph_solution = None
     if target_gates is not None and enabled_graph():
-        # Filter out MeasurementProcess instances that shouldn't be decomposed
-        decomposable_ops = [op for op in tape.operations if not isinstance(op, MeasurementProcess)]
+        # Filter out instances of ops that don't need to be decomposed
+        decomposable_ops = [op for op in tape.operations if not stopping_condition(op)]
```
Comment (mudit2812) [inline pennylane/devices/preprocess.py:377]: What about intermediate ops in the decomp graph that meet the stopping condition. Are those already handled automatically by the decomp graph?

File: pennylane/ops/op_math/controlled_ops.py [hunk 4]
```diff
@@ -251,12 +251,41 @@ def _controlled(self, wire):
...
+            **kwargs,
+        ): 1
+    }
+
+
+# pylint: disable=too-many-arguments
+@qml.register_condition(lambda num_target_wires, **_: num_target_wires > 2)
+@qml.register_resources(_to_general_c_qu_resource)
+def _to_general_c_qu(U, wires, control_wires, control_values, work_wires, work_wire_type, **_):
+    """A convert a ControlledQubitUnitary to a general Controlled(QubitUnitary) so that
```
Comment (mudit2812) [inline pennylane/ops/op_math/controlled_ops.py]:
```suggestion
    """Convert a ControlledQubitUnitary to a general Controlled(QubitUnitary) so that
```

File: pennylane/devices/preprocess.py (hunk 3 above)
Comment (astralcai) [reply pennylane/devices/preprocess.py:377]: The decomposition graph does not have early exit for intermediate ops because the decomposition graph fundamentally does not work with operator instances. This is just a partial optimization for operators in the top level circuit that don't need to be decomposed.

File: pennylane/math/decomposition.py
Comment (mudit2812) [inline pennylane/math/decomposition.py:1]: I'm having a bit of a hard time following the math here. Do you mind providing a short description of what is going on? Thanks in advance

**Review by mudit2812 — COMMENTED** [review]
Everything looks good. Just would like some details about the changes to the unitary matrix decomps

File: pennylane/math/decomposition.py [hunk 5]
```diff
@@ -44,18 +44,20 @@ def zyz_rotation_angles(U, return_global_phase=False):
     """
 
     U, alpha = math.convert_to_su2(U, return_global_phase=True)
-    # assume U = [[a, b], [c, d]], then here we take U[0, 1] as b
+    EPS = math.finfo(U.dtype).eps
```
Comment (JerryChen97) [inline pennylane/math/decomposition.py]: I feel its original location is actually better: it's only used once in the scope, so we did intentionally put it there before

Comment (astralcai) [reply pennylane/math/decomposition.py]: Sure

File: pennylane/math/decomposition.py [hunk 6]
```diff
@@ -80,22 +82,10 @@ def xyx_rotation_angles(U, return_global_phase=False):
...
-        math.isclose(math.sin(half_lam_plus_phi), math.zeros_like(half_lam_plus_phi)),
-        2 * math.arccos(math.clip(math.real(U[..., 1, 1]) / math.cos(half_lam_plus_phi), -1, 1)),
-        2 * math.arccos(math.clip(-math.imag(U[..., 0, 1]) / math.sin(half_lam_plus_phi), -1, 1)),
-    )
-
-    phi = math.squeeze(phi % (4 * np.pi))
-    theta = math.squeeze(theta % (4 * np.pi))
-    lam = math.squeeze(lam % (4 * np.pi))
+    C = math.cast_like(math.array([[1, -1], [1, 1]]) / np.sqrt(2), U)
+    U = math.einsum("mj, ...jk, kn -> ...mn", math.conjugate(C).T, U, C)
```
Comment (JerryChen97) [inline pennylane/math/decomposition.py:96]: Is `@` able to handle batch dim automatically?

File: pennylane/math/decomposition.py
Comment (astralcai) [reply pennylane/math/decomposition.py:1]: Added dev comments to explain the math.

File: pennylane/math/decomposition.py (hunk 5 above)
Comment (astralcai) [reply pennylane/math/decomposition.py]: Done

File: pennylane/math/decomposition.py (hunk 6 above)
Comment (astralcai) [reply pennylane/math/decomposition.py:96]: I've tried different solutions, either `torch` doesn't like it, or `jax` doesn't. This is the only one that works.

File: pennylane/math/decomposition.py
Comment (JerryChen97) [reply pennylane/math/decomposition.py:1]: So this original algo basically ends up a rot matrix in the middle e.g. 

$$
U = 
[ [C S]
  [S C] ]
$$
where C for cos(theta) S for sin(theta). I might have missed some minus sign somewhere.

In the dev notes we denoted U as `[[a, b], [c, d]]` though.
For some reason long ago (due to some incomplete implementation of atan2 for some interface, maybe tf, blended with differentiability issue), the developer chose to use arcsin for `b` to determine the rotation angle. However, apparently this cannot resolve the gauge issue, so they had to introduce an extra `where` to help get the correct branch.
But now it seems atan2 works very well! Didn't see any workflow breaks so I do believe this is a better choice. Note that equivalently one has `tan(theta) = sin/cos = b / a`. So it's a much better choice.

File: pennylane/math/decomposition.py (hunk 6 above)
Comment (JerryChen97) [reply pennylane/math/decomposition.py:96]: haha good to know. It's satisfying that both `torch` and `jax` supports einsum now.

File: pennylane/math/decomposition.py [hunk 7]
```diff
@@ -44,9 +44,10 @@ def zyz_rotation_angles(U, return_global_phase=False):
     """
 
     U, alpha = math.convert_to_su2(U, return_global_phase=True)
-    # assume U = [[a, b], [c, d]], then here we take U[0, 1] as b
```
Comment (JerryChen97) [inline pennylane/math/decomposition.py:47]: This note might be helpful actually. I think it even worth another sentence saying a is cos and b is sin therefore we take atan2 to extract the unique theta in the main branch [-pi, pi]

Comment (astralcai) [reply pennylane/math/decomposition.py:47]: Sure I'll add it back.

Comment (astralcai) [reply pennylane/math/decomposition.py:47]: [99fdb3d](https://github.com/PennyLaneAI/pennylane/pull/9036/commits/99fdb3d681036e45539c8fc2ef687d5e5999e864)

**Review by mudit2812 — APPROVED** [review]
Thanks for the dev comments, they were really helpful, and very easy to follow. Looks great!

**Review by JerryChen97 — APPROVED** [review]
Let's see if anything starts to break tmrw :pray: :cloud_with_lightning_and_rain:
